File: src/data/courses/dsa/string-hashing-problems/hashing-based-problems/longest-consecutive-sequence/code/python/longest_consecutive_sequence_sol2.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # Optimal: O(n) — Only count from sequence beginnings
        num_set = set(nums)
        max_len = 0
        for num in num_set:
            if num - 1 not in num_set:  # Sequence beginning
                cur, length = num, 1
                while cur + 1 in num_set: cur += 1; length += 1
                max_len = max(max_len, length)
        return max_len

File: src/data/courses/dsa/string-hashing-problems/hashing-based-problems/longest-consecutive-sequence/code/cpp/longest_consecutive_sequence_sol2.cpp
class Solution {
public:
    // Optimal: O(n) — Only count from sequence beginnings
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> numSet(nums.begin(), nums.end());
        int maxLen = 0;
        for (int num : numSet) {
            // Only start if this is a sequence BEGINNING
            if (!numSet.count(num - 1)) {
                int cur = num, len = 1;
                while (numSet.count(cur + 1)) { cur++; len++; }
                maxLen = max(maxLen, len);
            }
        }
        return maxLen;
    }
};